File: SchoolLms.Domain/Rooms.cs
namespace SchoolLms.Domain;

// ===========================================================================
//  Xonalar — docs/modules/students-parity.md §2.6 (R-1).
// ===========================================================================
//
//  NEGA SHU MODULDA
//  ----------------
//  `rooms` uchta modulga kerak: xonalar reyestri (§2.6), ombordagi xona
//  jihozlari (`docs/modules/warehouse.md` §3.2) va kelajakdagi jadval
//  moduli (`docs/SPEC.md` §3.3). Qoida: KIM BIRINCHI KELSA, O'SHA yaratadi va
//  boshqalar TAKRORLAMAYDI. Birinchi shu modul keldi.
//
//  SHAKL — AYNAN `SPEC.md` §3.3 DAGIDEK. Ustun ham, tur ham, DEFAULT ham
//  o'zgartirilmagan, aks holda keyingi ikki modul boshqa shaklni kutib qolardi.
//
//  NIMA QO'SHILMAYDI (§2.6.3, warehouse.md §3.2):
//    · `buildings` jadvali — maktab bitta binoda (bino nomi matn ustunda);
//    · `classes.home_room_id` va jadvalga ulanish — jadval modulining ishi;
//    · xona jihozlari — WareHouse moduli.
//
//  `SchoolClass.Room` (erkin matn) JOYIDA qoladi va uni o'qiydigan ekranlar
//  o'zgarmaydi; reyestr o'sha matnlardan BIR MARTA to'ldiriladi (seed).
// ===========================================================================

/// <summary><see cref="Room.Kind"/> qiymatlari (SPEC §3.3 dagi izoh).</summary>
public static class RoomKind
{
    public const string Classroom = "classroom";
    public const string Lab = "lab";
    public const string Gym = "gym";
    public const string Hall = "hall";

    public static readonly string[] All = [Classroom, Lab, Gym, Hall];

    public static bool IsValid(string? value) => value is not null && All.Contains(value);
}

/// <summary>Xona — band qilinadigan resurs (SPEC §3.3, R-1).</summary>
public class Room
{
    public Guid Id { get; set; } = Guid.NewGuid();

    /// <summary>Xona nomi/raqami ("101", "Sport zali"). Unikal.</summary>
    public string Name { get; set; } = string.Empty;

    /// <summary>Bino nomi (ixtiyoriy). Alohida `buildings` jadvali YO'Q.</summary>
    public string? Building { get; set; }

    public short? Floor { get; set; }

    /// <summary>Nechta o'quvchi sig'adi. SPEC §3.3: sukut 30.</summary>
    public short Capacity { get; set; } = 30;

    /// <summary>
    /// <see cref="RoomKind"/>. Baza CHECK bilan CHEKLAMAYDI — SPEC §3.3 da
    /// bu ro'yxat izoh, va WareHouse moduliga boshqa turdagi xona (ombor,
    /// kabinet) kerak bo'lishi mumkin; ro'yxatni migratsiya bilan kengaytirish
    /// o'sha modulning ishini bekorga sekinlashtirardi.
    /// </summary>
    public string Kind { get; set; } = RoomKind.Classroom;

    // =======================================================================
    //  §3.3 dan TASHQARI — R-1 slice'ining talabi (Batch C migratsiyasi).
    // =======================================================================

    /// <summary>
    /// Xona faolmi. <b>false = "ishlatilmaydi, lekin saqlanadi"</b>: yangi
    /// jadvalda va tanlovlarda ko'rinmaydi, lekin unga havola qilgan eski
    /// yozuvlar joyida qoladi.
    ///
    /// <para>
    /// <b>Nega kerak bo'ldi:</b> bugun reyestrda faqat O'CHIRISH bor, va
    /// sinf ko'rsatib turgan xonani o'chirib bo'lmaydi
    /// (<c>RoomTests.Sinf_korsatgan_xona_ochirilmaydi</c>) — ya'ni ta'mirga
    /// yopilgan xonani ro'yxatdan olib qo'yishning YO'LI YO'Q. Bu ustun
    /// <see cref="StudentStatus.IsActive"/> va
    /// <see cref="CertificateType.IsActive"/> bilan bir xil naqsh.
    /// </para>
    /// <para>
    /// Sukut <b>true</b> — mavjud har bir xona faol bo'lib qoladi, ya'ni
    /// bugungi ro'yxat va o'chirish qoidasi o'zgarmaydi.
    /// </para>
    /// </summary>
    public bool IsActive { get; set; } = true;
}

File: schoollms.client/src/api/services/rooms.ts
import { api, USE_MOCK } from '../client'

/* =========================================================================
 *  Xonalar reyestri — docs/modules/students-parity.md §2.6 (R-1).
 *
 *  Bu KATALOG: nom, bino, qavat, sig'im, tur. Xonani darsga biriktirish va
 *  to'qnashuvlarni tekshirish — jadval modulining ishi; xona jihozlari —
 *  WareHouse. Sinf kartochkasidagi "Xona" erkin matni O'ZGARMAYDI.
 *
 *  DTO nomlari backend bilan AYNAN bir xil (camelCase).
 * ========================================================================= */

/** Xona turlari — backenddagi `RoomKind` bilan bir xil ro'yxat. */
export type RoomKind = 'classroom' | 'lab' | 'gym' | 'hall'

export const roomKindLabels: Record<RoomKind, string> = {
  classroom: 'Sinf xonasi',
  lab: 'Laboratoriya',
  gym: 'Sport zali',
  hall: 'Yig‘ilish zali',
}

export const roomKindOptions: { value: RoomKind; label: string }[] = [
  { value: 'classroom', label: roomKindLabels.classroom },
  { value: 'lab', label: roomKindLabels.lab },
  { value: 'gym', label: roomKindLabels.gym },
  { value: 'hall', label: roomKindLabels.hall },
]

/** EduSchool ham ommaviy yaratishda shu chegarada to'xtatadi (§2.6.1). */
export const MAX_BULK_ROOMS = 50

export interface Room {
  id: string
  name: string
  building: string | null
  floor: number | null
  capacity: number
  kind: RoomKind
  /** Nechta sinf shu xonani ko'rsatgan. 0 dan katta bo'lsa o'chirib bo'lmaydi. */
  usedByClasses: number
  /**
   * Faolmi (Batch C qo'shimchasi, R-1 dan keyin). `false` — "ishlatilmaydi,
   * lekin saqlanadi": yangi jadval/sinfda ko'rinmaydi, uni ko'rsatgan eski
   * sinf esa joyida qoladi.
   */
  isActive: boolean
}

export interface SaveRoomInput {
  name: string
  building?: string | null
  floor?: number | null
  capacity?: number | null
  kind?: RoomKind
  /** Berilmasa — joyida qoladi (yaratishda sukut — faol). */
  isActive?: boolean
}

export interface RoomFilter {
  search?: string
  building?: string
  kind?: RoomKind
  /** Berilmasa — hammasi (faol ham, faolsiz ham). */
  isActive?: boolean
}

export interface BulkRoomsInput {
  count: number
  startFrom?: number | null
  prefix?: string
  building?: string | null
  floor?: number | null
  capacity?: number | null
  kind?: RoomKind
}

export interface BulkRoomsResult {
  created: Room[]
  /** Nomi band bo'lgani uchun yaratilmagan xonalar. */
  skipped: string[]
}

export async function getRooms(filter: RoomFilter = {}): Promise<Room[]> {
  if (USE_MOCK) return []
  const { data } = await api.get<Room[]>('/admin/rooms', { params: filter })
  return data
}

/** Binolar ro'yxati (filtr uchun). Alohida "binolar" katalogi yo'q — §2.6.3. */
export async function getRoomBuildings(): Promise<string[]> {
  if (USE_MOCK) return []
  const { data } = await api.get<string[]>('/admin/rooms/buildings')
  return data
}

export async function createRoom(input: SaveRoomInput): Promise<Room> {
  const { data } = await api.post<Room>('/admin/rooms', input)
  return data
}

export async function updateRoom(id: string, input: SaveRoomInput): Promise<Room> {
  const { data } = await api.put<Room>(`/admin/rooms/${id}`, input)
  return data
}

export async function deleteRoom(id: string): Promise<void> {
  await api.delete(`/admin/rooms/${id}`)
}

/** Ketma-ket raqamlangan N ta xona (EduSchool `rooms/multiple`). */
export async function createRoomsBulk(input: BulkRoomsInput): Promise<BulkRoomsResult> {
  const { data } = await api.post<BulkRoomsResult>('/admin/rooms/multiple', input)
  return data
}
